Introduce a package interface

elsh opened the pull request:
This PR introduces a new swiftinterface for package decls. 
- A .package.swiftinterface is now generated besides .private.swiftinterface and .swiftinterface (public). 
- The package interface file contains package decls as well as SPIs and public (and inlinable) APIs (superset of .private.swiftinterface).
- A new flag `-emit-package-module-interface-path` is added.

It also supports loading a module via interface if the importee and importer are in the same package. 
- Requires env `SWIFT_ENABLE_PACKAGE_INTERFACE_LOAD` or `-experimental-package-interface-load`   
- If opted-in, it now looks up a .package.swiftinterface before other interface files and checks if the importer is in the same package via the `-package-name` flag lookup. 
- If so, it builds a binary module from that package interface and loads that module. 
- If above fails or not opted-in, it falls back to the existing logic of module loading (looks up .private.swiftinterface and then public .swiftinterface) and shows appropriate diagnostics.
 
Resolves rdar://104617854

elsh committed:
Introduce a package interface.
It has an extension .package.swiftinterface and contains package decls
as well as SPIs and public/inlinable decls. When a module is loaded
from interface, it now looks up the package-name in the interface
and checks if the importer is in the same package. If so, it uses
that package interface found to load the module. If not, uses the existing
logic to load modules.

Resolves rdar://104617854

xymus commented on lib/Serialization/ScanningLoaders.cpp:
```diff
@@ -67,19 +67,27 @@ std::error_code SwiftModuleScanner::findModuleFilesInDirectory(
+  }
+
+  // If not in package, use the private interface file if exits.
+  if (fs.exists(InPath)) {
```
Shouldn't this be an `else if` to avoid overriding the package swiftinterface if it was picked?

elsh replied:
Oh, I meant the negation of that; will fix it.

nkcsgexi commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -596,17 +596,56 @@ std::string SerializedModuleBaseName::getName(file_types::ID fileTy) const {
   return std::string(result.str());
 }
 
+llvm::Optional<std::string> SerializedModuleLoaderBase::getPackageInterfacePathIfInSamePackage(llvm::vfs::FileSystem &fs, ASTContext &ctx) {
```
Already chatted with Ellie offline: at this point, it's better to put reading the package interface behind an env variable. After we ensure package interfaces are all healthy, we could lift the environment variable guard.

elsh replied:
Sg. Will add an env var as well as a flag.

nkcsgexi commented:
We also need to teach `swift-driver` to verify package-level textual interfaces.

xymus commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -596,17 +596,56 @@ std::string SerializedModuleBaseName::getName(file_types::ID fileTy) const {
+        StringRef current(args[I]), next(args[I + 1]);
+        if (current == "-package-name") {
+          result = next;
+          break;
```
It may be better not to `break` here and just keep looping. In case of flag duplication we usually pick the last one.

I would expect that when calling the compiler with `swift-frontend -package-name A -package-name B` we would already preserve B. That's from the default logic parsing options, we should try to reproduce it here. It's not applied much in the wild but it's pretty handy in tests with all the lit macros.

elsh replied:
will do

xymus commented on include/swift/AST/PrintOptions.h:
```diff
@@ -118,6 +118,12 @@ struct ShouldPrintChecker {
   virtual ~ShouldPrintChecker() = default;
 };
 
+enum class PrintInterfaceContentMode : uint8_t {
```
Could we use something shorter here? Since it's in a printing file we can probably cut some of the name as already being implied. How about just `ContentLevel` or `InterfaceLevel`?

elsh replied:
Could rename as InterfaceMode

elsh committed:
Allow loading package interface if in same package.
Add a new flag to enable package interface loading.
Use the last value of package-name in case of dupes.
Rename PrintInterfaceContentMode as InterfaceMode.
Update diagnostics.
Test package interface loading with various scenarios.
Test duplicate package-name.

elsh commented:
@swift-ci smoke test

artemcm commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -622,6 +622,12 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
     Opts.EnableAccessControl
       = A->getOption().matches(OPT_enable_access_control);
   }
+  if (auto A = Args.getLastArg(OPT_experimental_package_interface_load,
```
I think this should be:
```
hasArg(OPT_experimental_package_interface_load) {
```
?

elsh replied:
yup :) corrected

artemcm commented on lib/Frontend/ModuleInterfaceSupport.cpp:
```diff
@@ -621,7 +623,7 @@ class InheritedProtocolCollector {
 
     // Skip SPI extensions in the public interface.
-    if (!printOptions.PrintSPIs && extension->isSPI())
+    if (printOptions.InterfaceContentMode == InterfaceMode::Public && extension->isSPI())
```
We could probably put `.InterfaceContentMode == InterfaceMode::Public` into a helper method on `printOptions` to make this simpler, since it appears a lot.

artemcm commented on include/swift/Serialization/SerializedModuleLoader.h:
```diff
@@ -62,11 +62,17 @@ struct SerializedModuleBaseName {
+  /// Returns the .package.swiftinterface path if its package-name also applies to
+  /// the the importing module. Returns an empty optional otherwise.
+  llvm::Optional<std::string>
+  getPackageInterfacePathIfInSamePackage(llvm::vfs::FileSystem &fs,
```
Nit: It feels a bit overkill to pass in the whole Context here. I think we just need an optional package name?

elsh replied:
It checks if Ctx.LangOpts.EnablePackageInterfaceLoad as well. I can pass this boolean and package name as args though if preferred.

elsh committed:
Add an interface mode getter/setter in PrintOptions and ModuleInterfaceOptions
Update args parsing

elsh commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1103,14 +1103,14 @@ ERROR(module_not_testable,Fatal,
 WARNING(in_package_module_not_compiled_locally,none,
         "module %0 is in package %1 but was loaded from SDK; "
-        "modules of the same package should be built locally from source only: %2",
+        "modules of the same package should be built locally: %2",
```
Maybe it shouldn't be distributed. I'll leave the drag msg as is for now.

elsh committed:
Add a check for an empty package interface path
Update diags

elsh commented:
@swift-ci smoke test

elsh commented:
> We also need to teach `swift-driver` to verify package-level textual interfaces.

Will address it in a separate PR.

elsh committed:
Remove @objc CHECK as it only applies to Darwin

elsh commented:
@swift-ci smoke test linux windows

elsh commented:
@swift-ci smoke test

xymus commented on test/ModuleInterface/package_interface.swift:
```diff
@@ -0,0 +1,282 @@
+
+// RUN: %FileCheck -check-prefix=CHECK-LOAD-PUB %s < %t/load-pub.output
+// CHECK-LOAD-PUB: loaded module 'Bar'; source: '{{.*}}/Bar.swiftinterface', loaded: '{{.*}}/Bar-{{.*}}.swiftmodule'
+// CHECK-LOAD-PUB: error: module 'Bar' is in package 'barpkg' but was built from a non-package interface; modules of the same package can only be loaded if built from source or package interface: {{.*}}/Bar.swiftinterface
```
Could we make the test of the diagnostic a `-verify` check? It will ensure the error is on the import as expected.

elsh replied:
I'll update it in the followup PR that contains env var etc.

xymus commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -596,17 +596,62 @@ std::string SerializedModuleBaseName::getName(file_types::ID fileTy) const {
+  // Check if a package interface exists and if the package name applies to
+  // the importer module.
+  auto pkgPath = getPackageInterfacePathIfInSamePackage(fs, ctx).value_or("");
+  if (!pkgPath.empty() && fs.exists(pkgPath))
```
Isn't the `fs.exists(pkgPath)` check redundant as `getPackageInterfacePathIfInSamePackage` returns only existing paths already?

elsh replied:
Apparently it doesn't! The fs method is in llvm so will need to fix it there.

xymus commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -596,17 +596,62 @@ std::string SerializedModuleBaseName::getName(file_types::ID fileTy) const {
   return std::string(result.str());
 }
 
+llvm::Optional<std::string> SerializedModuleBaseName::getPackageInterfacePathIfInSamePackage(llvm::vfs::FileSystem &fs, ASTContext &ctx) const {
```
nitpick: This line looks unusually long, can we break it apart to try to keep in under the ~80 char per line limit?

xymus commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -632,6 +633,10 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
 
+  if (Args.hasArg(OPT_experimental_package_interface_load)) {
+    Opts.EnablePackageInterfaceLoad = true;
+  }
```
Should we add an env var as well to simplify qualifications? Something like:
```
else {
  Opts.EnablePackageInterfaceLoad  = ::getenv("SWIFT_PACKAGE_INTERFACE_LOAD")
}
```

elsh replied:
I have another PR that handles this but could add it here instead.

xymus reviewed (approved):
This looks good!

My last comments are mostly stylistic so they can wait to a later PR if this is ready to merge.

elsh committed:
Guard with env var SWIFT_ENABLE_PACKAGE_INTERFACE_LOAD
Update tests and formatting

elsh commented:
@swift-ci smoke test
